File: lib/clusterbuster/pod_files/uperf-client.pl
#!/usr/bin/perl

use Socket;
use POSIX;
use strict;
use Time::HiRes qw(gettimeofday usleep);
use Time::Piece;
use Sys::Hostname;
use File::Basename;
my ($dir) = $ENV{'BAK_CONFIGMAP'};
require "$dir/clientlib.pl";

my ($namespace, $container, $basetime, $baseoffset, $crtime,
    $exit_at_end, $synchost, $syncport, $loghost, $logport, $runtime, $ramp_time,
    $srvhost, $connect_port, @tests) = @ARGV;
my ($start_time, $data_start_time, $data_end_time, $elapsed_time, $end_time, $user, $sys, $cuser, $csys);
$start_time = xtime();
my ($processes) = 1;		# Not using multi-process here

$SIG{TERM} = sub { POSIX::_exit(0); };
$basetime += $baseoffset;
$crtime += $baseoffset;

my ($data_sent);
my ($mean_latency, $max_latency, $stdev_latency);

my $pass = 0;
my $ex = 0;
my $ex2 = 0;
my ($cfail) = 0;
my ($refused) = 0;
my ($pod) = hostname;
initialize_timing($basetime, $crtime, $synchost, $syncport, "$namespace:$pod:$container", $start_time);
$start_time = get_timing_parameter('start_time');

$SIG{TERM} = sub { POSIX::_exit(0); };
timestamp("Clusterbuster uperf client starting");

sub process_file($$%) {
    my ($infile, $outfile, %options) = @_;
    my ($contents);
    open IN, "<", "$infile" || die "Can't read $infile: $!\n";
    while (<IN>) {
	$contents .= $_;
    }
    close IN;
    foreach my $key (keys %options) {
	$contents =~ s/%{$key}/$options{$key}/g;
    }
    open OUT, ">", "$outfile" || die "Can't create $outfile: $!\n";
    print OUT $contents;
    close OUT || die "Can't close $outfile: $!\n";
}

my (%options) = (
    'srvhost' => $srvhost,
    'runtime' => 1,
);

process_file("$dir/uperf-mini.xml", "/tmp/fio-test.xml", %options);
# Ensure that uperf server is running before we try to do anything.
timestamp("Waiting for uperf server $srvhost:$connect_port to come online...");
system("bash", "-c", "until uperf -P $connect_port -m /tmp/fio-test.xml >/dev/null; do sleep 1; done");
timestamp("Connected to uperf server");

my ($counter) = 1;

sub compute_seconds($) {
    my ($value) = @_;
    if ($value =~ /^([[:digit:]]+(\.[[:digit:]]*)?)(ns|us|ms|s)$/) {
	my ($base) = $1;
	my ($modifier) = $3;
	if ($modifier eq 'us') {
	    return $base / 1000000.0;
	} elsif ($modifier eq 'ns') {
	    return $base / 1000000000.0;
	} elsif ($modifier eq 's') {
	    return $base / 1.0;
	} else {
	    return $base / 1000.0;
	}
    }
    return undef;
}

my (%results);
my (%cases);
my (@failed_cases) = ();

sub runit() {
    my ($ucpu0, $scpu0) = cputime();
    foreach my $test (@tests) {
	my ($test_type, $proto, $size, $nthr) = split(/, */, $test);
	my ($base_test_name) = "${proto}-${test_type}-${size}B-${nthr}i";
	my (%options) = (
	    'srvhost' => $srvhost,
	    'proto' => $proto,
	    'test_type' => $test_type,
	    'size' => $size,
	    'runtime' => $runtime + (2 * $ramp_time),
	    'nthr' => $nthr,
	    );
	my ($test_template) = "$dir/uperf-${test_type}.xml";
	my ($testfile) = "/tmp/fio-test.xml";
	process_file($test_template, $testfile, %options);
	my ($test_name) = sprintf('%04i-%s', $counter, $base_test_name);
	my (%metadata) = (
	    'protocol' => $proto,
	    'test_type' => $test_type,
	    'message_size' => $size,
	    'thread_count' => $nthr,
	    'test_name' => $test_name
	    );
	my ($failed) = 0;
	do_sync($synchost, $syncport, "$namespace:$pod:$container:$$:$test_name");
	timestamp("Running test $test_name");
	system("cat /tmp/fio-test.xml 1>&2");
	my ($job_start_time) = xtime();
	if (! defined $data_start_time) {
	    $data_start_time = $job_start_time;
	}
	open(RUN, "-|", "uperf", "-f", "-P", "$connect_port", '-m', '/tmp/fio-test.xml', '-R', '-a', '-i', '1', '-Tf') || die "Can't run uperf: $!\n";
	my ($start_time) = 0;
	my ($last_time) = 0;
	my ($last_nbytes) = 0;
	my ($last_nops) = 0;
	my ($ts_count) = 0;
	my (%case);
	my (@timeseries);
	my (%threads);
	my (%flowops);
	my (%summary) = (
	    'write' => {},
	    'read' => {},
	    'total' => {},
	    );
	my ($failure_message) = '';
	while (<RUN>) {
	    chomp;
	    timestamp($_);
	    if (/^timestamp_ms:([[:digit:].]+) +name:([[:alnum:]]+) +nr_bytes:([[:digit:]]+) +nr_ops:([[:digit:]]+)/) {
		my ($ts) = $1 / 1000.0;
		my ($name) = $2;
		my ($nbytes) = $3 + 0.0;
		my ($nops) = $4 + 0.0;
		# We only care about Txn2 and threads; the other transactions are start
		# and finish, and we want to ignore those
		if ($name eq 'Txn2') {
		    if ($start_time == 0) {
			$start_time = $ts;
			$last_time = $ts;
		    } else {
			my (%row) = (
			    'time' => $ts - $start_time,
			    'timedelta' => $ts - $last_time,
			    'bytes' => $nbytes - $last_nbytes,
			    'nops' => $nops - $last_nops,
			    );
			push @timeseries, \%row;
			$last_time = $ts;
			$last_nbytes = $nbytes;
			$last_nops = $nops;
		    }
		} elsif ($name =~ /^Thr([[:digit:]])+/) {
		    my (%row) = (
			'time' => $ts - $start_time,
			'bytes' => $nbytes,
			'nops' => $nops,
			);
		    $threads{$name} = %row;
		}
	    } elsif (/^(Txn1|write|read)[ \t]/) {
		my ($op, $count, $avg, $cpu, $max, $min) = split;
		if ($op eq 'Txn1') {
		    $op = 'total';
		}
		$summary{$op}{'time_avg'} = compute_seconds($avg);
		$summary{$op}{'time_max'} = compute_seconds($max);
		$summary{$op}{'time_min'} = compute_seconds($min);
	    } elsif (/^[*][*] Error/) {
		$failure_message = $_;
		$failed = 1;
		timestamp("Test case $test_name failed!");
		push @failed_cases, $test_name;
	    } elsif (/^[*]/) {
		timestamp($_);
	    } elsif (/WARNING: Errors/ && ! $failed) {
		$failure_message = $_;
		$failed = 1;
		timestamp("Test case $test_name failed!");
		push @failed_cases, $test_name;
	    }
	}
	close(RUN);
	$data_end_time = xtime();
	$summary{'raw_elapsed_time'} = $last_time - $start_time;
	$summary{'raw_nbytes'} = $last_nbytes;
	$summary{'raw_nops'} = $last_nops;
	if ($summary{'raw_elapsed_time'} > 0) {
	    $summary{'raw_avg_ops_sec'} = $summary{'raw_nops'} / $summary{'raw_elapsed_time'};
	    $summary{'raw_avg_bytes_sec'} = $summary{'raw_nbytes'} / $summary{'raw_elapsed_time'};
	}
	$summary{'nbytes'} = 0;
	$summary{'nops'} = 0;
	$summary{'elapsed_time'} = 0;
	$summary{'avg_bytes_sec'} = 0;
	$summary{'avg_ops_sec'} = 0;
	my ($ops_sec_sum) = 0;
	my ($ops_sec_sq_sum) = 0;
	my ($bytes_sec_sum) = 0;
	my ($bytes_sec_sq_sum) = 0;
	my ($stdev_counter) = 0;
	map {
	    $summary{'nbytes'} += $$_{'bytes'};
	    $summary{'nops'} += $$_{'nops'};
	    $summary{'elapsed_time'} += $$_{'timedelta'};
	    my ($ops_sec) = $$_{'nops'} / $$_{'timedelta'};
	    my ($bytes_sec) = $$_{'bytes'} / $$_{'timedelta'};
	    $ops_sec_sum += $ops_sec;
	    $ops_sec_sq_sum += $ops_sec * $ops_sec;
	    $bytes_sec_sum += $bytes_sec;
	    $bytes_sec_sq_sum += $bytes_sec * $bytes_sec;
	    $stdev_counter++;
	} grep { ($summary{'raw_elapsed_time'} < 10 ||
		  ($$_{'time'} >= $ramp_time &&
		   $$_{'time'} < $summary{'raw_elapsed_time'} - $ramp_time)) } @timeseries;
	if ($summary{'elapsed_time'} > 0) {
	    $summary{'avg_bytes_sec'} = $summary{'nbytes'} / $summary{'elapsed_time'};
	    $summary{'avg_ops_sec'} = $summary{'nops'} / $summary{'elapsed_time'};
	    $summary{'bytes_sec_sq_sum'} = $bytes_sec_sq_sum;
	    $summary{'bytes_sec_sum'} = $bytes_sec_sum;
	    $summary{'ops_sec_sq_sum'} = $ops_sec_sq_sum;
	    $summary{'ops_sec_sum'} = $ops_sec_sum;
	    if ($stdev_counter >= 2) {
		$summary{'stdev_bytes_sec'} = (($bytes_sec_sq_sum / $stdev_counter) - ($summary{'avg_bytes_sec'} ** 2)) ** 0.5;
		$summary{'stdev_ops_sec'} = (($ops_sec_sq_sum / $stdev_counter) - ($summary{'avg_ops_sec'} ** 2)) ** 0.5;
	    } else {
		$summary{'stdev_bytes_sec'} = 0;
		$summary{'stdev_ops_sec'} = 0;
	    }
	}
	$summary{'job_start'} = $job_start_time;
	$summary{'job_end'} = $data_end_time;
	$case{'status'}{'condition'} = $failed ? 'FAIL' : 'PASS';
	$case{'status'}{'message'} = $failure_message;
	$case{'metadata'} = \%metadata;
	$case{'summary'} = \%summary;
	$case{'timeseries'} = \@timeseries;
	$elapsed_time += $summary{'elapsed_time'};
	$counter++;
	$cases{$test_name} = \%case;
    }
    $results{'results'} = \%cases;
    $results{'failed'} = \@failed_cases;

    my ($ucpu1, $scpu1) = cputime();
    $ucpu1 -= $ucpu0;
    $scpu1 -= $scpu0;

    my ($results) = print_json_report($namespace, $pod, $container, $$, $data_start_time,
				      $data_end_time, $elapsed_time, $ucpu1, $scpu1, \%results);
    timestamp("Done");
    if ($syncport) {
	do_sync($synchost, $syncport, $results);
    }
    if ($logport > 0) {
	do_sync($loghost, $logport, $results);
    }
}

my (%pids) = ();
for (my $i = 0; $i < $processes; $i++) {
    my $child;
    if (($child = fork()) == 0) {
	runit();
	exit(0);
    } else {
	$pids{$child} = 1;
    }
}
while (%pids) {
    my ($child) = wait();
    if ($child == -1) {
	finish($exit_at_end);
    } elsif (defined $pids{$child}) {
	if ($?) {
	    timestamp("Pid $child returned status $?!");
	    finish($exit_at_end, $?, $namespace, $pod, $container, $synchost, $syncport, $child);
	}
	delete $pids{$child};
    }
}

finish($exit_at_end);

File: lib/clusterbuster/pod_files/uperf-client.py
#!/usr/bin/env python3

import os
import subprocess
import re
import json
import traceback

from clusterbuster_pod_client import clusterbuster_pod_client


class uperf_client(clusterbuster_pod_client):
    """
    uperf test for clusterbuster
    """

    def __init__(self):
        super().__init__()
        self.runtime = int(self._args[0])
        self.ramp_time = int(self._args[1])
        self.srvhost = self.resolve_host(self._args[2])
        self.connect_port = int(self._args[3])
        self.tests = self._args[4:]
        self.podfile_dir = os.environ.get('PODFILE_DIR', '.')
        self.process_file(os.path.join(self.podfile_dir, "uperf-mini.xml"), "/tmp/uperf-test.xml", {'srvhost': self.srvhost, 'runtime': 1})
        self.timestamp(f"Waiting for uperf server {self.srvhost}:{self.connect_port} to come online...")
        subprocess.run(f'until uperf -P "{self.connect_port}" -m /tmp/uperf-test.xml; do sleep 1; done', shell=True)
        self.timestamp("Connected to uperf server")

    def compute_seconds_uperf(self, value: str):
        # Specific to uperf, which defaults to milliseconds
        m = re.match(r'([0-9]+(\.[0-9]*)?)(ns|us|ms|s)$', value)
        if m:
            base = float(m.group(1))
            modifier = m.group(3)
            if modifier == 'us':
                return base / 1000000
            elif modifier == 'ns':
                return base / 1000000000
            elif modifier == 's':
                return base
            else:
                return base / 1000
        else:
            return None

    def process_file(self, infile: str, outfile: str, options: dict):
        with open(infile, 'r') as indata:
            contents = indata.read()
        for key, sub in options.items():
            contents = contents.replace(f"%{{{key}}}", str(sub))
        with open(outfile, 'w') as outdata:
            outdata.write(contents)

    def runit(self, process: int):
        ucpu, scpu = self.cputimes()
        counter = 1
        data_start_time = None
        failed_cases = []
        results = dict()
        cases = dict()
        elapsed_time = 0
        for test in self.tests:
            [test_type, proto, size, nthr] = re.split(r'[,\s]+', test)
            size = int(size)
            nthr = int(nthr)
            base_test_name = f"{proto}-{test_type}-{size}B-{nthr}i"
            options = {
                'srvhost': self.srvhost,
                'proto': proto,
                'test_type': test_type,
                'size': size,
                'runtime': self.runtime + (2 * self.ramp_time),
                'nthr': nthr
                }
            test_template = os.path.join(self.podfile_dir, f"uperf-{test_type}.xml")
            testfile = "/tmp/uperf-test.xml"
            self.process_file(test_template, testfile, options)
            test_name = '%04i-%s' % (counter, base_test_name)
            metadata = {
                'protocol': proto,
                'test_type': test_type,
                'message_size': size,
                'thread_count': nthr,
                'test_name': test_name
                }
            failed = False
            self.sync_to_controller(self.idname(test_name))
            self.timestamp(f"Running test {test_name}")
            with open(testfile, 'r') as f:
                self.timestamp(f.read())
            job_start_time = self.adjusted_time()
            if data_start_time is None:
                data_start_time = job_start_time
            with subprocess.Popen(["uperf", "-f", "-P", str(self.connect_port), '-m', '/tmp/uperf-test.xml', '-R', '-a', '-i', '1', '-Tf'], stdout = subprocess.PIPE) as run:
                first_time = 0
                last_time = 0
                last_nbytes = 0
                last_nops = 0
                timeseries = []
                threads = dict()
                summary = {
                    'write': {},
                    'read': {},
                    'total': {}
                    }
                failure_message = ''
                line = run.stdout.readline().decode('ascii')
                while line:
                    line = line.strip()
                    m = re.match(r'timestamp_ms:([0-9.]+) +name:([0-9a-zA-Z]+) +nr_bytes:([0-9]+) +nr_ops:([0-9]+)', line)
                    if m:
                        ts = float(m.group(1)) / 1000.0
                        name = m.group(2)
                        nbytes = int(m.group(3))
                        nops = int(m.group(4))
                        # We only care about Txn2 and threads; the other transactions are start
                        # and finish, and we want to ignore those
                        if name == 'Txn2':
                            row = {
                                'time': ts - first_time,
                                'timedelta': ts - last_time,
                                'nbytes': nbytes - last_nbytes,
                                'nops': nops - last_nops
                                }
                            timeseries.append(row)
                            last_time = ts
                            last_nbytes = nbytes
                            last_nops = nops
                        elif re.match(r'Thr([0-9]+)', name):
                            row = {
                                'time': ts - first_time,
                                'nbytes': nbytes,
                                'nops': nops
                                }
                            threads[name] = row
                    elif re.match(r'(Txn1|write|read)\s', line):
                        [op, count, avg, cpu, maximum, minimum] = line.split()
                        if op == 'Txn1':
                            op = 'total'
                        summary[op]['time_avg'] = self.compute_seconds_uperf(avg)
                        summary[op]['time_max'] = self.compute_seconds_uperf(maximum)
                        summary[op]['time_min'] = self.compute_seconds_uperf(minimum)
                    elif line.startswith('** Error') or (line.startswith('WARNING: Errors') and not failed):
                        failure_message = line
                        failed = True
                        self.timestamp(f"Test case {test_name} failed!")
                        failed_cases.append(test_name)
                    elif line.startswith('*'):
                        self.timestamp(line)
                    else:
                        pass
                    line = run.stdout.readline().decode('ascii')
                status = run.poll()
                if failed or status != 0:
                    self.timestamp(f"Uperf failed: {status}")
                    return 1
            job_end_time = self.adjusted_time()
            data_end_time = job_end_time
            summary['raw_elapsed_time'] = last_time - first_time
            summary['raw_nbytes'] = last_nbytes
            summary['raw_nops'] = last_nops
            if summary['raw_elapsed_time'] > 0:
                summary['raw_avg_ops_sec'] = summary['raw_nops'] / summary['raw_elapsed_time']
                summary['raw_avg_bytes_sec'] = summary['raw_nbytes'] / summary['raw_elapsed_time']
            summary['nbytes'] = 0
            summary['nops'] = 0
            summary['elapsed_time'] = 0
            summary['avg_bytes_sec'] = 0
            summary['avg_ops_sec'] = 0
            ops_sec_sum = 0
            ops_sec_sq_sum = 0
            bytes_sec_sum = 0
            bytes_sec_sq_sum = 0
            stdev_counter = 0
            for point in timeseries:
                if summary['raw_elapsed_time'] < 10 or (point['time'] >= self.ramp_time and
                                                        point['time'] < summary['raw_elapsed_time'] - self.ramp_time):
                    summary['nbytes'] += point['nbytes']
                    summary['nops'] += point['nops']
                    summary['elapsed_time'] += point['timedelta']
                    ops_sec = point['nops'] / point['timedelta']
                    bytes_sec = point['nbytes'] / point['timedelta']
                    ops_sec_sum += ops_sec
                    ops_sec_sq_sum += ops_sec ** 2
                    bytes_sec_sum += bytes_sec
                    bytes_sec_sq_sum += bytes_sec ** 2
                    stdev_counter += 1
            if summary['elapsed_time'] > 0:
                summary['avg_bytes_sec'] = summary['nbytes'] / summary['elapsed_time']
                summary['avg_ops_sec'] = summary['nops'] / summary['elapsed_time']
                summary['bytes_sec_sq_sum'] = bytes_sec_sq_sum
                summary['bytes_sec_sum'] = bytes_sec_sum
                summary['ops_sec_sq_sum'] = ops_sec_sq_sum
                summary['ops_sec_sum'] = ops_sec_sum
                if stdev_counter >= 2:
                    summary['stdev_bytes_sec'] = ((bytes_sec_sq_sum / stdev_counter) - (summary['avg_bytes_sec'] ** 2)) ** 0.5
                    summary['stdev_ops_sec'] = ((ops_sec_sq_sum / stdev_counter) - (summary['avg_ops_sec'] ** 2)) ** 0.5
                else:
                    summary['stdev_bytes_sec'] = 0
                    summary['stdev_ops_sec'] = 0
            summary['job_start_time'] = job_start_time
            summary['job_end_time'] = job_end_time
            case = {
                'metadata': metadata,
                'summary': summary,
                'timeseries': timeseries,
                'status': {
                    'message': failure_message
                    }
                }
            if failed:
                case['status']['condition'] = 'FAIL'
            else:
                case['status']['condition'] = 'PASS'
            elapsed_time += summary['elapsed_time']
            counter = counter + 1
            cases[test_name] = case
        results['results'] = cases
        results['failed'] = failed_cases
        ucpu, scpu = self.cputimes(ucpu, scpu)
        self.report_results(data_start_time, data_end_time, elapsed_time, ucpu, scpu, results)


uperf_client().run_workload()
